force-app/main/default/classes/DMLOperations.cls

```diff
@@ -17,204 +17,337 @@
+   * If an Account with the given name already exists, update the description to "Updated Account".
+   * If no such Account exists, it creates a new one and sets the description to "New Account".
+   * @param accountName The name of the Account to be upserted.
+   * @return The Account that was upserted.
+   */
+  public static Account upsertAccount(String accountName) {
+    // Query for the Account
+    List<Account> accounts = [SELECT Id FROM Account WHERE Name = :accountName];
+    Account account;
+
+    // if the Account exists, update the description
+    // otherwise, create a new Account and set the description
+    if (!accounts.isEmpty()) {
+      account = accounts.get(0);
+      account.Description = 'Updated Account';
+    } else {
+      account = new Account(Name = accountName);
+      account.Description = 'New Account';
     }
-
-    /**
-     * Question 6
-     * Update the Account's Name and Industry given an Account Id.
-     * Resource: https://developer.salesforce.com/docs/atlas.en-us.soql_sosl.meta/soql_sosl/sforce_api_calls_soql.htm
-     * @param accountId The Id of the account to be updated.
-     * @param newName The new name for the account.
-     * @param newIndustry The new industry for the account.
-     */
-    public static void updateAccountFields(Id accountId, String newName, String newIndustry) {
-        // Query the account
-        // [SELECT Id FROM Account WHERE Id = :accountId]
-
-        // Update the Account with the given Id with the new name and industry
+    // Upsert the Account
+    upsert account;
+    return account;
+  }
+
+  /**
+   * Question 10
+   * Manage the association between Contacts and Accounts
+   * Ensure that for each Contact in the provided list, there is a corresponding Account Name that matches the Contact's last name.
+   * If such an Account does not exist, the method creates it; if it does, the method ensures the Contact is associated with it. (Hint: use previous upsertAccount method - this may cause bad practices but for learning purposes it's okay.)
+   * Example Contact.LastName = 'Doe' Link to Account.Name = 'Doe'
+   * Example Contact.LastName = 'Jane' Link to Account.Name = 'Jane'
+   * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
+   */
+  public static void upsertAccountsWithContacts(
+    List<Contact> contactsToUpsert
+  ) {
+    // Loop through each Contact
+
+    // Extract the Account Name from Contact's LastName
+
+    // Get the account based on the name or create one if it doesn't exist
+
+    // Link the Contact to its Account using the Account Id
+
+    // Upsert the Contacts
+
+    for (Contact contact : contactsToUpsert) {
+      String accountName = contact.LastName;
+      List<Account> accounts = [
+        SELECT Id
+        FROM Account
+        WHERE Name = :accountName
+      ];
+      if (accounts.isEmpty()) {
+        Account newAccount = new Account(Name = accountName);
+        insert newAccount;
```

Comment:
Try to avoid doing DML inside loops except for the one mentioned in the comment for practice